File: Leetcode/tree/559_MaximumDepthOfN-aryTree/Solution.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/
class Solution {
public:
    int maxDepth(Node* root) {
        /*/ //recursive
        if(root==NULL) return 0;
        int res = 0;
        for(auto child: root->children) res = max(res, maxDepth(child));
        return 1+res; /*/
        if(root==nullptr) return 0;
        queue<Node*> q;
        q.push(root);
        int res = 0;
        while(!q.empty()){
            int size = q.size();
            for(int i = 0; i<size; i++){
                Node* cur = q.front(); q.pop();
                for(auto child: cur->children) q.push(child);
            }
            res++;
        }
        return res;
    }
};

File: Leetcode/tree/559_MaximumDepthOfN-aryTree/Solution.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val,List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/
class Solution {
    public int maxDepth(Node root) {
        if(root==null) return 0;
        int res = 0;
        for(Node child: root.children)  res = Math.max(res, maxDepth(child));
        return 1+res;
        /*/
        if(root==null) return 0;
        Queue<Node> queue = new LinkedList<>();
        queue.offer(root);
        int res = 0;
        while(!queue.isEmpty()){
            int size = queue.size();
            for(int i = 0; i<size; i++){
                Node current = queue.poll();
                for(Node child: current.children) queue.offer(child);
            }
            res++;
        }
        return res;/*/
    }
}
